File: library-core/src/main/java/com/mikepenz/fastadapter/IAdapterNotifier.java
package com.mikepenz.fastadapter;

/**
 * Created by mikepenz on 28.10.17.
 */

public interface IAdapterNotifier {

    IAdapterNotifier DEFAULT = new IAdapterNotifier() {
        @Override
        public boolean notify(FastAdapter fastAdapter, int newItemsCount, int previousItemsCount, int itemsBeforeThisAdapter) {
            //now properly notify the adapter about the changes
            if (newItemsCount > previousItemsCount) {
                if (previousItemsCount > 0) {
                    fastAdapter.notifyAdapterItemRangeChanged(itemsBeforeThisAdapter, previousItemsCount);
                }
                fastAdapter.notifyAdapterItemRangeInserted(itemsBeforeThisAdapter + previousItemsCount, newItemsCount - previousItemsCount);
            } else if (newItemsCount > 0) {
                fastAdapter.notifyAdapterItemRangeChanged(itemsBeforeThisAdapter, newItemsCount);
                if (newItemsCount < previousItemsCount) {
                    fastAdapter.notifyAdapterItemRangeRemoved(itemsBeforeThisAdapter + newItemsCount, previousItemsCount - newItemsCount);
                }
            } else if (newItemsCount == 0) {
                fastAdapter.notifyAdapterItemRangeRemoved(itemsBeforeThisAdapter, previousItemsCount);
            } else {
                //this condition should practically never happen
                fastAdapter.notifyAdapterDataSetChanged();
            }
            return false;
        }
    };

    IAdapterNotifier LEGACY_DEFAULT = new IAdapterNotifier() {
        @Override
        public boolean notify(FastAdapter fastAdapter, int newItemsCount, int previousItemsCount, int itemsBeforeThisAdapter) {
            //now properly notify the adapter about the changes
            if (newItemsCount > previousItemsCount) {
                if (previousItemsCount > 0) {
                    fastAdapter.notifyAdapterItemRangeChanged(itemsBeforeThisAdapter, previousItemsCount);
                }
                fastAdapter.notifyAdapterItemRangeInserted(itemsBeforeThisAdapter + previousItemsCount, newItemsCount - previousItemsCount);
            } else if (newItemsCount > 0 && newItemsCount < previousItemsCount) {
                fastAdapter.notifyAdapterItemRangeChanged(itemsBeforeThisAdapter, newItemsCount);
                fastAdapter.notifyAdapterItemRangeRemoved(itemsBeforeThisAdapter + newItemsCount, previousItemsCount - newItemsCount);
            } else if (newItemsCount == 0) {
                fastAdapter.notifyAdapterItemRangeRemoved(itemsBeforeThisAdapter, previousItemsCount);
            } else {
                //this condition should practically never happen
                fastAdapter.notifyAdapterDataSetChanged();
            }
            return false;
        }
    };

    boolean notify(FastAdapter fastAdapter, int newItemsCount, int previousItemsCount, int itemsBeforeThisAdapter);
}

File: library-core/src/main/java/com/mikepenz/fastadapter/IAdapterNotifier.kt
package com.mikepenz.fastadapter

/**
 * Created by mikepenz on 28.10.17.
 */

interface IAdapterNotifier {

    fun notify(fastAdapter: FastAdapter<*>, newItemsCount: Int, previousItemsCount: Int, itemsBeforeThisAdapter: Int): Boolean

    companion object {
        @JvmField
        val DEFAULT: IAdapterNotifier = object : IAdapterNotifier {
            override fun notify(fastAdapter: FastAdapter<*>, newItemsCount: Int, previousItemsCount: Int, itemsBeforeThisAdapter: Int): Boolean {
                //now properly notify the adapter about the changes
                when {
                    newItemsCount > previousItemsCount -> {
                        if (previousItemsCount > 0) {
                            fastAdapter.notifyAdapterItemRangeChanged(itemsBeforeThisAdapter, previousItemsCount)
                        }
                        fastAdapter.notifyAdapterItemRangeInserted(itemsBeforeThisAdapter + previousItemsCount, newItemsCount - previousItemsCount)
                    }
                    newItemsCount > 0 -> {
                        fastAdapter.notifyAdapterItemRangeChanged(itemsBeforeThisAdapter, newItemsCount)
                        if (newItemsCount < previousItemsCount) {
                            fastAdapter.notifyAdapterItemRangeRemoved(itemsBeforeThisAdapter + newItemsCount, previousItemsCount - newItemsCount)
                        }
                    }
                    newItemsCount == 0 -> fastAdapter.notifyAdapterItemRangeRemoved(itemsBeforeThisAdapter, previousItemsCount)
                    //this condition practically should never happen
                    else -> fastAdapter.notifyAdapterDataSetChanged()
                }
                return false
            }
        }

        @JvmField
        val LEGACY_DEFAULT: IAdapterNotifier = object : IAdapterNotifier {
            override fun notify(fastAdapter: FastAdapter<*>, newItemsCount: Int, previousItemsCount: Int, itemsBeforeThisAdapter: Int): Boolean {
                //now properly notify the adapter about the changes
                when {
                    newItemsCount > previousItemsCount -> {
                        if (previousItemsCount > 0) {
                            fastAdapter.notifyAdapterItemRangeChanged(itemsBeforeThisAdapter, previousItemsCount)
                        }
                        fastAdapter.notifyAdapterItemRangeInserted(itemsBeforeThisAdapter + previousItemsCount, newItemsCount - previousItemsCount)
                    }
                    newItemsCount in 1 until previousItemsCount -> {
                        fastAdapter.notifyAdapterItemRangeChanged(itemsBeforeThisAdapter, newItemsCount)
                        fastAdapter.notifyAdapterItemRangeRemoved(itemsBeforeThisAdapter + newItemsCount, previousItemsCount - newItemsCount)
                    }
                    newItemsCount == 0 -> fastAdapter.notifyAdapterItemRangeRemoved(itemsBeforeThisAdapter, previousItemsCount)
                    //this condition practically should never happen
                    else -> fastAdapter.notifyAdapterDataSetChanged()
                }
                return false
            }
        }
    }
}
